### PR TITLE
Updates the TS patch

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

- We weren't compatible with TS 4.2 (beta, but still)
- The TS patch script was a monstrosity

Fixes https://github.com/yarnpkg/berry/issues/1869 (https://github.com/arcanis/TypeScript/pull/3)
Fixes https://github.com/yarnpkg/berry/issues/1932 (https://github.com/arcanis/TypeScript/pull/4)
Fixes https://github.com/yarnpkg/berry/issues/2223 (https://github.com/arcanis/TypeScript/pull/5)
Fixes https://github.com/yarnpkg/berry/issues/2384
Closes https://github.com/yarnpkg/berry/pull/2215

**How did you fix it?**
I rewrote the patch script in JS, took the opportunity to add validation into it, and added listing for 4.2. It's much faster too!

I also removed support for pre-4.0 TS versions (in order to keep the number of patches low), making this a major change.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-compat/extra/typescript/patch-f347bb4623a8370ac25e4dda78c964e9.diff [hunk 1]
```diff
@@ -0,0 +1,2335 @@
+warning: CRLF will be replaced by LF in /lib/lib.d.ts.
```
Comment (merceyz) [inline packages/plugin-compat/extra/typescript/patch-f347bb4623a8370ac25e4dda78c964e9.diff:1]: We should probably silence / replace these, unsure if the code `patch-commit` uses also produces these :thinking:

Comment (arcanis) [reply packages/plugin-compat/extra/typescript/patch-f347bb4623a8370ac25e4dda78c964e9.diff]: It was a mistake, stderr wasn't meant to be piped into the file if the exit code is 0.

**Comment by pravi** [comment]
Will there be a new release with these changes soon? This breaks the native debian package installation (uses yarn 2.4 and node-modules linker).

**Comment by pravi** [comment]
Workaround, lock typescript to 4.1.5 in resolutions.

**Comment by arcanis** [comment]
https://github.com/yarnpkg/berry/issues/2514#issuecomment-785122658

**Pull request merged**